manager/scheduler/scheduler.go

```diff
@@ -427,30 +446,97 @@ func (s *Scheduler) taskFitNode(ctx context.Context, t *api.Task, nodeID string)
 	return &newT
 }
 
-// scheduleTask schedules a single task.
-func (s *Scheduler) scheduleTask(ctx context.Context, t *api.Task) *api.Task {
+// scheduleTaskGroup schedules a batch of tasks that are part of the same
+// service and share the same version of the spec.
+func (s *Scheduler) scheduleTaskGroup(ctx context.Context, taskGroup map[string]*api.Task, schedulingDecisions map[string]schedulingDecision) {
+	// Pick at task at random from taskGroup to use for constraint
+	// evaluation. It doesn't matter which one we pick because all the
+	// tasks in the group are equal in terms of the fields the constraint
+	// filters consider.
+	var t *api.Task
+	for _, t = range taskGroup {
+		break
+	}
+
 	s.pipeline.SetTask(t)
-	n, _ := s.nodeHeap.findMin(s.pipeline.Process, s.scanAllNodes)
-	if n == nil {
-		log.G(ctx).WithField("task.id", t.ID).Debug("No suitable node available for task")
-		return nil
+
+	nodeLess := func(a *NodeInfo, b *NodeInfo) bool {
+		sameServiceTasksOnA := 0
+		sameServiceTasksOnB := 0
+
+		// TODO(aaronl): NodeInfo should keep a map of serviceID->[]task,
+		// so that checking how many tasks from a given service are
+		// running on the node is cheap.
+		for _, aTask := range a.Tasks {
+			if aTask.ServiceID == t.ServiceID {
+				sameServiceTasksOnA++
+			}
+		}
+		for _, bTask := range b.Tasks {
+			if bTask.ServiceID == t.ServiceID {
+				sameServiceTasksOnB++
+			}
+		}
+
+		if sameServiceTasksOnA < sameServiceTasksOnB {
+			return true
+		}
+		if sameServiceTasksOnB > sameServiceTasksOnA {
+			return false
+		}
+
+		// Total number of tasks breaks ties.
+		return len(a.Tasks) < len(b.Tasks)
 	}
 
-	log.G(ctx).WithField("task.id", t.ID).Debugf("Assigning to node %s", n.ID)
-	newT := *t
-	newT.NodeID = n.ID
-	newT.Status = api.TaskStatus{
-		State:     api.TaskStateAssigned,
-		Timestamp: ptypes.MustTimestampProto(time.Now()),
-		Message:   "scheduler assigned task to node",
+	nodes := s.nodeHeap.findBestNodes(len(taskGroup), s.pipeline.Process, nodeLess)
+	if len(nodes) == 0 {
+		for _, t := range taskGroup {
+			log.G(ctx).WithField("task.id", t.ID).Debug("no suitable node available for task")
+			s.enqueue(t)
+		}
+		return
 	}
-	s.allTasks[t.ID] = &newT
 
-	nodeInfo, err := s.nodeHeap.nodeInfo(n.ID)
-	if err == nil && nodeInfo.addTask(&newT) {
-		s.nodeHeap.updateNode(nodeInfo)
+	nodeIter := 0
+	for taskID, t := range taskGroup {
```

Comment:
This loop should keep assigning tasks to the current "best" node until it is no longer the best node or doesn't meet the constraints, then move on to the next node (see https://github.com/docker/swarmkit/pull/1446#discussion_r76359670).